### PR TITLE
Language server protocol (LSP)

### LINKED ISSUE #67: Pre-proposal: incorporation of jupyter(lab)-lsp
## Pre-Proposal Checklist

- [x] Briefly outlines the proposal
- [ ] Suggests a review team (optional)
- [x] Declares why it should be a JEP (See the “JEP / Not-a-JEP Rubric” below.)
- [ ] Identify a Shepherd to see the process through. (Shepherds are assigned on a round-robin basis from a set of interested engaged volunteers).
- [ ] Decide if it’s a JEP according to the criteria listed above. The Shepherd decides if the JEP criteria have been met.

## Pre-Proposal
The maintainers of [jupyter(lab)-lsp](https://github.com/krassowski/jupyterlab-lsp) would like to propose its incorporation as an official sub-project of Project Jupyter. We feel this would benefit Jupyter users through better discoverability of advanced interactive computing features, supported by the [Language Server Protocol](https://langserver.org) (LSP), but otherwise missing in a user's Jupyter experience.

The key component of the repository, [@krassowski/jupyterlab-lsp](https://www.npmjs.com/package/@krassowski/jupyterlab-lsp), offers Jupyter users an expanded subset of features described by the LSP as an extension to JupyterLab. These features include refinements of existing Jupyter interactive computing features, such as completion and introspection, as well as new Jupyter features such as linting, reference linking, and symbol renaming. It is supported by [jupyter-lsp](https://pypi.org/project/jupyter-lsp/), a Language Server- and Jupyter Client-agnostic extension of the Jupyter Notebook Server (for the `0.x` line) and Jupyter Server (for the `1.x`). We will discuss the architecture and engineering process of maintaining these components at greater length, leveraging a good deal of the user and developer [documentation](https://jupyterlab-lsp.readthedocs.io/en/latest/?badge=latest).

Much like [Jupyter Kernel Messaging](https://jupyter-client.readthedocs.io/en/stable/messaging.html), LSP provides a language-agnostic, JSON-compatible description for multiple clients to integrate with any number of language implementations. Unlike Kernel Messaging, the focus is on precise definition of the many facets of static analysis and code transformation, with nearly four times the number of messages of the Jupyter specification. We will discuss the opportunities and challenges of this complexity for users and maintainers of Jupyter Clients, Kernels, and related tools.

### Branding

We invite discussion on an appropriate Jupyter branding/organization approach, ranging from:
- a single move of the `@krassowski/jupyterlab-lsp` monorepo to `@jupyter/lsp`
- a new, Jupyter co-branded GitHub organization (e.g. `jupyter-lsp`)

### Kernel integration

In order for `jupyterlab-lsp` to be able to connect to the appropriate _Language Server_ in tandem with a _Kernel_, it requires that kernels properly implemenet [KernelInfo](https://jupyter-client.readthedocs.io/en/stable/messaging.html#kernel-info) from version `5.0` of Jupyter Messaging specification, in particular the field `language_info` with the following fields present:
- `mimetype`
- `file_extension`
- `name`

Therefore, **no changes to the specification are required**. Though some future specification refinements could be discussed as separate JEPs.

In continuous integration, we keep under test _Kernels_ and _Language Servers_ for all three of the Jupyter-name forming languages:
- Julia: IJulia using LanguageServer.jl,
- Python: IPython using pyls or jedi-language-server,
- R: IRKernel using R languageserver

Other kernels also confirmed to work straightaway (Bash) or after minimal user configuration (Scala) in presence of appropriate _Language Servers_. When accounting for language servers that only support the file editor but not notebooks, a total of [13 language servers](https://jupyterlab-lsp.readthedocs.io/en/latest/Language%20Servers.html) is known to work with the jupyterlab-lsp.

Julia-Python-R-opt

Integration with alternative frontends might be possible in the future; as a proof-of-concept we were able to activate [most of the LSP functionality](https://github.com/krassowski/jupyterlab-lsp/pull/465) in the jupyterlab-classic (a prototype re-implementation of the classic interface using JupyterLab components).

Interestingly, the extension was adopted as a basis for other notable JupyterLab extensions, and it was used to prototype interesting novel concepts in [academic context](https://github.com/wala/graph4code/blob/master/docs/figures/demo_v2.mp4).

### Magics and polyglot support

We also support the magics in IPython, allowing for polyglot experience when using cell magics. For the detailed discussions on magics integration please refer to [our discussion on *code overrides*](https://github.com/krassowski/jupyterlab-lsp/issues/347) (which is our term for generalisation of magics, shell expressions and pinfo `?`/`??` syntax).

rpy2-demo

### Outlook towards notebook-specific IDE features

The team envisions a support for the notebook specific IDE features to be included in the future. This might involve upstream changes to the LSP specification, or custom jupyter-specific extension. Please see [our community discussion](https://github.com/krassowski/jupyterlab-lsp/issues/467) for more details.

### Minimal implementation details

- `jupyterlab-lsp` implements integration with JupyterLab and its default CodeMirror editor; the design allows for incorporation of additional editors in the future, but large portions of code are CodeMirror 5-specific; it communicates with `jupyter-lsp` to send/recieve messages from language servers and with JupyterLab to get information about currrent kernel
- `jupyter-lsp` uses `jupyter_server`, `tornado`, sockets and IO streams for communication with servers and with frontend; it is not aware of a kernel's existence
- we use GitHub CI running `pytest`, `jest`, and `robot` (Selenium) test; the integration/acceptance tests for the bulk of the frontend testing; we test on Windows, Linux and MacOS, supporting thee Python versions at a time for `jupyter-lsp` and doing tests with actual language servers

### Previous discussions:

Please refer to:
- https://github.com/jupyterlab/team-compass/issues/67 for the discussion held within the JupyterLab team.
- https://github.com/jupyter/enhancement-proposals/pull/26 for previous (stale) JEP on LSP adoption; this proposal superseds that JEP; importantly jupyter(lab)-lsp does not require any specific editor, whereas the previous JEP specifically mentions Monaco.

### PR DESCRIPTION
**Pull request by krassowski**
As discussed in the pre-proposal and in this JEP we would like to propose adoption of the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) to enhance coding assistance in Jupyter notebooks (and other editors used by Jupyter frontends). In addition to rather broad JEP (we do not propose a specific implementation as a final standard) we would also like to convert the [current repository](https://github.com/krassowski/jupyterlab-lsp) where prototyping and exploration of this topic happens into a GitHub organization following Jupyter governance, contribution guidelines and branding to foster collaboration in this area (as discussed in the pre-proposal).

Closes #67 (pre-proposal) and closes #26 (supersedes).

@blink1073 is acting as the shepherd for the proposal.

Pre-requisites:

1. [x] **Create a GitHub issue**  in the Jupyter Enhancement Proposals repo that
   1. [x] Briefly outlines the proposal
   2. [ ] Suggests a review team (optional)
   3. [x] Declares why it should be a JEP (See the “JEP / Not-a-JEP Rubric” below.)
2. [x] **Identify a Shepherd** to see the process through. (Shepherds are assigned
   on a round-robin basis from a set of interested engaged volunteers).
3. [x] **Decide if it's a JEP** according to the criteria listed above. The Shepherd decides if the JEP criteria have been met.
4. [x] **Used the new JEP markdown template**

The pre-proposal received several (what I assume are) positive reactions, but no comments in the 6 months since it was opened.

- [x] @afshin
- [x] @blink1073
- [x] @Carreau
- [x] @damianavila
- [x] @ellisonbg
- [x] @fperez
- [ ] @ivanov
- [x] @jasongrout
- [ ] @jhamrick
- [x] @minrk
- [x] @mpacer
- [ ] @parente
- [ ] @rgbkrk
- [ ] @Ruv7
- [x] @SylvainCorlay
- [ ] @takluyver
- [x] @willingc

### CODE REVIEW

**Comment by krassowski** [comment]
What should I do to move this forward?

**Comment by blink1073** [comment]
Opening the RFC phase, pinging @lresende @williamstein @hbcarlos  @jtpio @ellisonbg @rgbkrk @Zsailer @ 
TylerLeonhardt @MSeal @blois @choldgraf @captainsafia @westurner who have all engaged on this topic.

**Comment by blink1073** [comment]
Eight folks have given a thumbs up on the PR text, let's move this into the Final Comment Period (FCP) until 26 August.

File: 72-language-server-protocol/language-server-protocol.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+Jupyter ecosystem by leveraging the established
+[Language Server Protocol](https://microsoft.github.io/language-server-protocol/) (LSP), with a good
+overview on the [community knowledge site](https://langserver.org). We would like to propose its
+incorporation as an official sub-project of Project Jupyter. We feel this would benefit Jupyter
+users through better discoverability of advanced interactive computing features, supported by the
+(LSP), but otherwise missing in a user's Jupyter experience. While our repository currently features
+a working implementation, the proposal is not tied to it (beyond a proposal for migration of the
+repository to a Jupyter-managed GitHub organization) but rather aimed to guide the process of
+formalizing and evolving the way of integrating Jupyter with LSP in general.
+
```
[the remark this answers is not in the document]
Comment (westurner) [reply 72-language-server-protocol/language-server-protocol.md:22]: How will LSP interact with - is it called / issue # ? - multikernel notebooks with multiple languages? Apologies if this is already elsewhere addressed.

[the remark this answers is not in the document]
Comment (krassowski) [reply 72-language-server-protocol/language-server-protocol.md:22]: Yes, this is mentioned in the JEP under "polyglot" term. One discussion of this issue is in https://github.com/krassowski/jupyterlab-lsp/issues/282

[the remark this answers is not in the document]
Comment (westurner) [reply 72-language-server-protocol/language-server-protocol.md:22]: > How to implement LSP for a multi-language kernel (SoS)? https://github.com/krassowski/jupyterlab-lsp/issues/282

**Review by damianavila — APPROVED** [review]

**Review by SylvainCorlay — APPROVED** [review]

**Review by Carreau — APPROVED** [review]

**Review by fperez — APPROVED** [review]
Thanks so much @krassowski for this detailed document and your LSP work! 

I'm very enthusiastic about this - I recall discussions with MS folks early on when LSP was just getting started, about how well its ideas tied into the Jupyter model, but we hadn't come to closer integration. I fully support having this, and I think once it lands, it's the kind of foundation that can lead to a lot of interesting exploration, new tooling, etc.

Big 👍 from me, thx again!

**Review by mpacer — APPROVED** [review]
I appreciate the summary and the deep dives in the comments. Interoperability with well-established standards like LSP seems worthwhile and exciting. Thanks all for pushing this through!

**Review by afshin — APPROVED** [review]

**Review by choldgraf — APPROVED** [review]
I agree w/ what @mpacer says :-) I think it's a good idea!

**Review by jasongrout — APPROVED** [review]
This looks great. Thank you for all of your hard work on this project and JEP!

**Review by willingc — APPROVED** [review]

**Review by MSeal — APPROVED** [review]

**Review by minrk — APPROVED** [review]

**Review by damianavila — APPROVED** [review]

**Comment by blink1073** [comment]
Merging since it has been 4 weeks and all votes have been in favor.  Thanks @krassowski!

**Pull request merged**